source/common/upstream/load_balancer_impl.cc

```diff
@@ -337,47 +337,146 @@ const HostVector& ZoneAwareLoadBalancerBase::tryChooseLocalLocalityHosts(const H
+  // been added/removed, etc. but this is quite a bit more complicated and will require some
+  // additional addMemberUpdateCb parameters to track efficiently. The other downside of a full
+  // recompute is that time complexity is O(n * log n), so we will need to do better at delta
+  // tracking to scale.
+  priority_set.addMemberUpdateCb(
+      [this](uint32_t priority, const HostVector&, const HostVector&) { refresh(priority); });
 }
 
-HostConstSharedPtr RoundRobinLoadBalancer::chooseHost(LoadBalancerContext*) {
-  const HostVector& hosts_to_use = hostsToUse();
-  if (hosts_to_use.empty()) {
-    return nullptr;
+void RoundRobinLoadBalancer::refresh(uint32_t priority) {
+  const auto add_hosts_source = [this](HostsSource source, const HostVector& hosts) {
+    bool weighted = false;
+
+    for (const auto& host : hosts) {
+      if (host->weight() != hosts[0]->weight()) {
+        weighted = true;
+        break;
+      }
+    }
+
+    // Nuke existing scheduler if it exists.
+    scheduler_[source] = Scheduler{};
+    scheduler_[source].weighted_ = weighted;
+    if (weighted) {
+      // Populate scheduler with host list.
+      for (const auto& host : hosts) {
+        // We use a fixed weight here. While the weight may change without
+        // notification, this will only be stale until this host is next picked,
+        // at which point it is reinserted into the EdfScheduler with its new
+        // weight in chooseHost().
+        scheduler_[source].edf_.add(host->weight(), host);
+      }
+    }
+  };
+  // Populate EdfSchedulers for each valid HostsSource value for the host set
+  // at this priority.
+  const auto& host_set = priority_set_.hostSetsPerPriority()[priority];
+  add_hosts_source(HostsSource(priority, HostsSource::SourceType::AllHosts), host_set->hosts());
+  add_hosts_source(HostsSource(priority, HostsSource::SourceType::HealthyHosts),
+                   host_set->healthyHosts());
+  for (uint32_t locality_index = 0;
+       locality_index < host_set->healthyHostsPerLocality().get().size(); ++locality_index) {
+    add_hosts_source(
+        HostsSource(priority, HostsSource::SourceType::LocalityHealthyHosts, locality_index),
+        host_set->healthyHostsPerLocality().get()[locality_index]);
   }
+}
 
-  return hosts_to_use[rr_index_++ % hosts_to_use.size()];
+HostConstSharedPtr RoundRobinLoadBalancer::chooseHost(LoadBalancerContext*) {
+  const HostsSource hosts_source = hostSourceToUse();
+  auto scheduler_it = scheduler_.find(hosts_source);
+  // We should always have a scheduler for any return value from
+  // hostSourceToUse() via the construction in refresh();
+  ASSERT(scheduler_it != scheduler_.end());
+  auto& scheduler = scheduler_it->second;
+  if (scheduler.weighted_) {
+    auto host = scheduler.edf_.pick();
+    // We should always succeed if weighted, since when we compute the scheduler
+    // in refresh() above, any empty host vector will be treated as unweighted.
+    ASSERT(host != nullptr);
```

Comment:
I think this is the only bit I'm twitchy about because if there's any bugs there's not great error handling.  If there's only one host and it is expired, we're good?

Reply:
We don't expire until we rebuild the full host list, at which point we're also rebuilding the schedule. Comment added.